File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private Deque<Integer> preorderDeque;
    private HashMap<Integer, Integer> inorderMap;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        // init the preorder deque
        preorderDeque = new LinkedList<>();
        for (int val : preorder) {
            preorderDeque.addLast(val);
        }

        // hashmap for val : i wrt to inorder array
        inorderMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }

        // recursion returns root node
        return helper(0, inorder.length - 1);
    }

    private TreeNode helper(int l, int r) {
        // Base case: impossible
        if (l > r) {
            return null;
        }

        // first element is always the root in preorder
        int rootVal = preorderDeque.removeFirst();
        TreeNode root = new TreeNode(rootVal);

        // get m from inorder array.
        int m = inorderMap.get(rootVal);

        // recursion
        root.left = helper(l, m - 1);
        root.right = helper(m + 1, r);

        return root;
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # first node of preorder is always the root
        # root element in inorder traversal will divide array into two parts, left and right.
        # recursively run for both leaf nodes.
        preorder = deque(preorder)

        # hashmap for val : index in the inorder table
        mp = {val : i for i, val in enumerate(inorder)}

        def helper(l, r):
            # base case:
            if l > r: #impossible
                return None
            
            # get the first element in preorder, which is the root.
            val = preorder.popleft()
            root = TreeNode(val)

            # find index of root val in inorder array using map.
            m = mp[val]

            # recursion:
            root.left = helper(l, m-1)
            root.right = helper(m+1, r)

            return root

        return helper(0, len(inorder) - 1) # returns root.
